### DISCUSSION TITLE
Implicit casts

### DISCUSSION THREAD

**Discussion by soerenwolfers** [duckdb/duckdb#11805] [Polls]
Personally, I prefer errors over implicit casts.

Just some examples of surprising results based on recent issues that I read:
```sql
SELECT 1 || 0, ARRAY_VALUE(1) || ARRAY_VALUE(0), CONCAT_WS(', ', [1, 0])
```
```
┌──────────┬────────────────────────────────────┬────────────────────────────────────────┐
│ (1 || 0) │ (array_value(1) || array_value(0)) │ concat_ws(', ', main.list_value(1, 0)) │
│ varchar  │              varchar               │                varchar                 │
├──────────┼────────────────────────────────────┼────────────────────────────────────────┤
│ 10       │ [1][0]                             │ [1, 0]                                 │
└──────────┴────────────────────────────────────┴────────────────────────────────────────┘
```

I understand it's a tradeoff between, essentially, the demands of interactive work, where one wants to just quickly get things working, and production code, where one wants confidence that things will work as expected or break loudly. 

I'd be interested how the duckdb userbase is split on this.

**Comment by Tishj** [comment]
https://github.com/duckdb/duckdb/pull/10194

**Reply by Tishj** [reply]
This doesn't really constitute as an implicit cast in the eyes of the system.
The `concat` function / `concat_ws` function / `||` operator take ANY as argument type, and cast the arguments to VARCHAR

**Reply by Tishj** [reply]
I do agree this might be a little unexpected behavior, seeing as we have this setting, with this default:
```c++
	//! Use old implicit casting style (i.e. allow everything to be implicitly casted to VARCHAR)
	bool old_implicit_casting = false;
```

One might expect this to error unless this is set to true

**Comment by tjmcewan** [comment]
The first example on the [postgres compat page](https://duckdb.org/docs/stable/sql/dialect/postgresql_compatibility#implicit-casting-on-equality-checks), oh boy.  `'1.1' = 1` O_o

Just flagging it would be great to have a setting to disable implicit casting altogether. 🙏🏻
